### DISCUSSION TITLE
Run the Pretrained Model without a Validation Set or Test Set

### DISCUSSION THREAD

**Discussion by 20Bolin** [bio-ontology-research-group/mowl#80] [General]
### Describe the bug

Hi, I tried to implement mowl on my ontology for the embedding. From what I see in the documentation, the validation and test ontology are just optional. 

![image](https://github.com/user-attachments/assets/ffcaf838-591d-4ec3-bc75-81827aeb22d8)

And I don't have validation set or test set at the moment, so I tried to run the model with my own training dataset.

However, it raised an error stating that "AttributeError: Validation dataset is None.", as shown in the screenshot below.

![image](https://github.com/user-attachments/assets/35d6b3b4-1acd-4b2f-b3db-58b82a0f16ca)
![image](https://github.com/user-attachments/assets/74848593-2df1-4e3c-842c-992935a06750)

I wonder whether it is impossible to run the model without validation ontology or test ontology. If so, how can I create those validation set or test set from the ontology I have at the moment? I have some ontology in the owl format similar to this:
     

    

    

    <owl:Class rdf:about="http://www.opengis.net/ont/geosparql#Feature">
        <rdfs:subClassOf rdf:resource="http://www.opengis.net/ont/geosparql#SpatialObject"/>
        <owl:disjointWith rdf:resource="http://www.opengis.net/ont/geosparql#Geometry"/>
        <rdfs:isDefinedBy rdf:resource="http://www.opengis.net/ont/geosparql#"/>
        <rdfs:isDefinedBy rdf:resource="http://www.opengis.net/spec/geosparql/1.0/req/core/feature-class"/>
        <rdfs:isDefinedBy rdf:resource="http://www.opengis.net/spec/geosparql/1.1/req/core/feature-class"/>
        <skos:definition xml:lang="en">A discrete spatial phenomenon in a universe of discourse.</skos:definition>
        <skos:example rdf:resource="http://www.opengis.net/spec/geosparql/1.1/specification.html#C.1.1.2.1"/>
        <skos:example rdf:resource="http://www.opengis.net/spec/geosparql/1.1/specification.html#C.1.1.2.2"/>
        <skos:example rdf:resource="http://www.opengis.net/spec/geosparql/1.1/specification.html#C.1.1.2.3"/>
        <skos:example rdf:resource="http://www.opengis.net/spec/geosparql/1.1/specification.html#C.1.1.2.4"/>
        <skos:example rdf:resource="http://www.opengis.net/spec/geosparql/1.1/specification.html#C.1.1.2.5"/>
        <skos:example rdf:resource="http://www.opengis.net/spec/geosparql/1.1/specification.html#C.1.1.2.6"/>
        <skos:example rdf:resource="http://www.opengis.net/spec/geosparql/1.1/specification.html#C.1.1.2.7"/>
        <skos:example rdf:resource="http://www.opengis.net/spec/geosparql/1.1/specification.html#C.1.1.2.8"/>
        <skos:example rdf:resource="http://www.opengis.net/spec/geosparql/1.1/specification.html#C.1.1.2.9"/>
        <skos:example rdf:resource="http://www.opengis.net/spec/geosparql/1.1/specification.html#C.1.1.3.2"/>
        <skos:example rdf:resource="http://www.opengis.net/spec/geosparql/1.1/specification.html#C.1.1.3.3"/>
        <skos:example rdf:resource="http://www.opengis.net/spec/geosparql/1.1/specification.html#C.1.2.2"/>
        <skos:example rdf:resource="http://www.opengis.net/spec/geosparql/1.1/specification.html#C.1.2.3"/>
        <skos:example rdf:resource="http://www.opengis.net/spec/geosparql/1.1/specification.html#C.1.2.4"/>
        <skos:note xml:lang="en">A Feature represents a uniquely identifiable phenomenon, for example a river or an apple. While such phenomena (and therefore the Features used to represent them) are bounded, their boundaries may be crisp (e.g., the declared boundaries of a state), vague (e.g., the delineation of a valley versus its neighboring mountains), and change with time (e.g., a storm front). While discrete in nature, Features may be created from continuous observations, such as an isochrone that determines the region that can be reached by ambulance within 5 minutes.</skos:note>
        <skos:prefLabel xml:lang="en">Feature</skos:prefLabel>
    </owl:Class>

### How to reproduce

```
from tqdm import trange, tqdm
import torch as th
import torch.nn as nn
from mowl.datasets.base import PathDataset, Dataset, RemoteDataset, OWLClasses
from mowl.models.elembeddings.examples.model_ppi import ELEmPPI
from mowl.projection.factory import projector_factory

#Build a subclass of PathDataset to implement a customized evaluation_class
class CustomDataset(PathDataset):
    @property
    def evaluation_classes(self):
        """Classes that are used in evaluation
        """

        if self._evaluation_classes is None:
            gis = set()
            for owl_name, owl_cls in self.classes.as_dict.items():
                if "http://www.opengis" in owl_name:
                    gis.add(owl_cls)
            self._evaluation_classes = OWLClasses(gis), OWLClasses(gis)

        return self._evaluation_classes

ds = CustomDataset(ontology_path="opengis.owl")
dataset = ds

model = ELEmPPI(dataset,
                embed_dim=30,
                margin=0.1,
                reg_norm=1,
                learning_rate=0.001,
                epochs=20,
                batch_size=4096,
                model_filepath=r"C:\Users\Steven\mOWL",
                device='cpu')

# Set the number of individuals
model.module.ind_embed = nn.Embedding(num_embeddings=len(dataset.classes), embedding_dim=30)

# Training
model.train()
```

### Environment

Windows 10
Python version 3.11.9
JDK version 11.0.22

**Comment by ferzcam** [comment]
Hi. Since you are using `mowl.models.elembeddings.examples.model_ppi.ELEmPPI` directly, validation loss is computed as part of the training loop. I would suggest to modify the `ELEmPPI` training loop yourself and remove/adapt the validation loss part. That would be removing/adapting the code starting at: https://github.com/bio-ontology-research-group/mowl/blob/466494d9c1d6db664a358452f6508772f8fbc172/mowl/models/elembeddings/examples/model_ppi.py#L54
until:
https://github.com/bio-ontology-research-group/mowl/blob/466494d9c1d6db664a358452f6508772f8fbc172/mowl/models/elembeddings/examples/model_ppi.py#L64

Let me know if this helps.

**Comment by 20Bolin** [comment]
> Hi. Since you are using `mowl.models.elembeddings.examples.model_ppi.ELEmPPI` directly, validation loss is computed as part of the training loop. I would suggest to modify the `ELEmPPI` training loop yourself and remove/adapt the validation loss part. That would be removing/adapting the code starting at:
> 
> https://github.com/bio-ontology-research-group/mowl/blob/466494d9c1d6db664a358452f6508772f8fbc172/mowl/models/elembeddings/examples/model_ppi.py#L54
> 
> 
> until:
> https://github.com/bio-ontology-research-group/mowl/blob/466494d9c1d6db664a358452f6508772f8fbc172/mowl/models/elembeddings/examples/model_ppi.py#L64
> 
> Let me know if this helps.

Hi! At the beginning, the error was raise in line 26 and said the dataset doesn't have evaluation_classes and this must be implemented in a child class. So I removed everything related to the variable "prots" and it caused the AttributeError I showed above.
https://github.com/bio-ontology-research-group/mowl/blob/466494d9c1d6db664a358452f6508772f8fbc172/mowl/models/elembeddings/examples/model_ppi.py#L25
https://github.com/bio-ontology-research-group/mowl/blob/466494d9c1d6db664a358452f6508772f8fbc172/mowl/models/elembeddings/examples/model_ppi.py#L26

This time I simply removed line from 54 till 64, the training seems to work with any error messages. Even if I keep prots, no issues about the dataset.evaluation_classes were raised, which is good, but I don't quite understand why.
Now another problem is that the training can be executed, but no saved model can be found in the model path I designated, which is quite weird. How can I fix this problem?


On the other hand, I don't think it makes enough sense to train without validation and test. So I also tried to build validation set and test set from the ontology I have right now. I tried to use Owlready2, but I don't know how to do it properly. The code I used is shown below:
```
from owlready2 import get_ontology, Thing, Ontology
import random

ontology = get_ontology("opengis.owl").load()
classes = list(ontology.classes())

# Shuffle the classes
random.shuffle(classes)

# Define the split ratios
train_ratio = 0.6
val_ratio = 0.2
test_ratio = 0.2

# Calculate split indices
train_end = int(train_ratio * len(classes))
val_end = train_end + int(val_ratio * len(classes))

# Split the classes
train_classes = classes[:train_end]
val_classes = classes[train_end:val_end]
test_classes = classes[val_end:]

# Initialize the split ontology with a unique IRI
train_ontology = get_ontology("urn:example:train_ontology")
val_ontology = get_ontology("urn:example:val_ontology")
test_ontology = get_ontology("urn:example:test_ontology")

# Function to add classes to a new ontology
def add_classes_to_ontology(classes, new_ontology):
    with new_ontology:
        for cls in classes:
            new_cls = type(cls.name, (Thing,), {})
            new_cls.is_a = cls.is_a

# Add classes to the respective ontologies
add_classes_to_ontology(train_classes, train_ontology)
add_classes_to_ontology(val_classes, val_ontology)
add_classes_to_ontology(test_classes, test_ontology)

# Save the ontologies to OWL files
train_ontology.save(file="opengis_train.owl", format="rdfxml")
val_ontology.save(file="opengis_val.owl", format="rdfxml")
test_ontology.save(file="opengis_test.owl", format="rdfxml")
```

But I don't think it's a proper way to do it, because in the example valid.owl and test.owl, there is a structured schema with object properties and classes while in the output of my method, I don't have it. Could you give me some advice on the train-test-split on ontology data? I asked it here but maybe it would be better to quote and raise another issue since it is a bit different from the previous question.

**Comment by 20Bolin** [comment]
My ontology is also in owl and I want the validation and test also in owl and have a structured schema like:

```
    <!-- 
    ///////////////////////////////////////////////////////////////////////////////////////
    //
    // Object Properties
    //
    ///////////////////////////////////////////////////////////////////////////////////////
     -->

    


    <!-- http://has_function -->

    <owl:ObjectProperty rdf:about="http://has_function"/>
    


    <!-- http://has_label -->

    <owl:ObjectProperty rdf:about="http://has_label"/>
    


    <!-- http://interacts_with -->

    <owl:ObjectProperty rdf:about="http://interacts_with"/>
    


    <!-- http://purl.obolibrary.org/obo/BFO_0000050 -->

    <owl:ObjectProperty rdf:about="http://purl.obolibrary.org/obo/BFO_0000050">
        <owl:inverseOf rdf:resource="http://purl.obolibrary.org/obo/BFO_0000051"/>
        <rdf:type rdf:resource="http://www.w3.org/2002/07/owl#TransitiveProperty"/>
        <oboInOwl:hasDbXref>BFO:0000050</oboInOwl:hasDbXref>
        <oboInOwl:hasOBONamespace>external</oboInOwl:hasOBONamespace>
        <oboInOwl:id>part_of</oboInOwl:id>
        <oboInOwl:shorthand>part_of</oboInOwl:shorthand>
        <rdfs:label>part of</rdfs:label>
    </owl:ObjectProperty>

    <!-- 
    ///////////////////////////////////////////////////////////////////////////////////////
    //
    // Classes
    //
    ///////////////////////////////////////////////////////////////////////////////////////
     -->

    


    <!-- http://4932.Q0010 -->

    <owl:Class rdf:about="http://4932.Q0010">
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://interacts_with"/>
                <owl:someValuesFrom rdf:resource="http://4932.Q0017"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://interacts_with"/>
                <owl:someValuesFrom rdf:resource="http://4932.Q0032"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://interacts_with"/>
                <owl:someValuesFrom rdf:resource="http://4932.Q0092"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://interacts_with"/>
                <owl:someValuesFrom rdf:resource="http://4932.Q0142"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://interacts_with"/>
                <owl:someValuesFrom rdf:resource="http://4932.Q0182"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://interacts_with"/>
                <owl:someValuesFrom rdf:resource="http://4932.Q0297"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://interacts_with"/>
                <owl:someValuesFrom rdf:resource="http://4932.YDL114W"/>
            </owl:Restriction>
        </rdfs:subClassOf>
    </owl:Class>
    


    <!-- http://4932.Q0017 -->

    <owl:Class rdf:about="http://4932.Q0017">
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://interacts_with"/>
                <owl:someValuesFrom rdf:resource="http://4932.Q0010"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://interacts_with"/>
                <owl:someValuesFrom rdf:resource="http://4932.Q0032"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://interacts_with"/>
                <owl:someValuesFrom rdf:resource="http://4932.Q0092"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://interacts_with"/>
                <owl:someValuesFrom rdf:resource="http://4932.Q0142"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://interacts_with"/>
                <owl:someValuesFrom rdf:resource="http://4932.Q0143"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://interacts_with"/>
                <owl:someValuesFrom rdf:resource="http://4932.Q0182"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://interacts_with"/>
                <owl:someValuesFrom rdf:resource="http://4932.Q0297"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://interacts_with"/>
                <owl:someValuesFrom rdf:resource="http://4932.YDL114W"/>
            </owl:Restriction>
        </rdfs:subClassOf>
    </owl:Class>
```

**Comment by ferzcam** [comment]
Hello. In order to create a dataset, you can follow examples like:
https://github.com/bio-ontology-research-group/mowl/blob/466494d9c1d6db664a358452f6508772f8fbc172/extra/create_dataset.py where we create the PPI benchmark. In this example, the split the protein-protein interactions for train, validation and testing and insert the training set into the Gene Ontology. 

Also, every dataset has different evaluation entities. For example here you can find two cases https://github.com/bio-ontology-research-group/mowl/blob/466494d9c1d6db664a358452f6508772f8fbc172/experiments/datasets.py . The first one is a subsumption dataset where axioms $C \sqsubseteq D$ are to be evaluated and $C,D$ are named classes. The other one is for the PPI benchmark where axioms $p_1 \sqsubseteq \exists interacts\_with . p_2$ are to be evaluated and $p_1,p_2$ are classes representing proteins. For your use case, you might want to generate your own dataset. 

Let me know if this clarifies the process to create your dataset.

**Reply by 20Bolin** [reply]
Hi, from what I see in create_dataset.py, the ontology file xxx.owl, the data file and the annotation file are a must in the input to create the dataset. What is contained in the data file here? Is it instance data? And, I only have an ontology file. I don't have the annotation file and datafile yet. How can I create these files for my ontology? (Maybe I can query the instance data of the ontology from the knowledge graph if that's the data file but I have no idea how to make an annotation file in .sgd.gaf.short for owl ontology. Also what does the .short extension mean here?) In genetics it seems there are pretty well-defined GO annotations, GO knowledge base, GO annotation file(GAF) and it seems that these ontology are by default with annotations, which makes them work compatibly with mowl. But what about ontology in other domains, like mine in geo-information?

In addition, I note that there are indeed annotation properties in my owl ontology. Are these the so-called annotation that should be extracted and construct the annotation file?

Thank you for your patience in advance! I am quite new to ontology engineering so please excuse me if I ask some naive questions!

**Reply by ferzcam** [reply]
Hi. The dataset you create depends on the use case. For example, in the PPI benchmark, we obtain protein--protein interactions from some database and split those interaction as training, validation and testing. Training interactions are added to the Gene Ontology, validation and testing interactions are added to new empty ontologies. 
Another case is the subsumption prediction task. In this task, we extracted all axioms $C \sqsubseteq D$ from GO or FoodOn. We split those axioms into training validation and testing. We added the training axioms back to the remaining of GO or FoodOn and then we put the validation and testing axioms in new empty ontologies. The .short or .gaf files here are specific to the PPI case, maybe for your case you do not need those files but you might have some sources of data that you can use to generate training, validation and testing sets.

By annotations, in the PPI case, we refer to protein--protein interactions statements. I understand this might be confused with annotation properties in ontologies. Sorry for the confusion. 

Please let me know if this clarifies the process to create the dataset. I am happy to help! Maybe if you can provide some details on the task you are working on, the statements you want to predict or how you want to use ontology embeddings I can also provide a better help.

**Reply by 20Bolin** [reply]
Hi, thx for your reply! What I want to do right now is related to knowledge graph question answering. There is a chatbot and a question answering system to support the chatbot. The question answering system is based on a big knowledge graph with billions of triples. And it is related to cadastral data. Since the question answering system is also backed by an LLM, we plan to use ontology embedding and match in the prompt of the LLM to generate sparql queries.
Because there are a lot of triples in the knowledge graph and due to token limitation, it is not practical to embed the whole knowledge graph and put the embedding into the prompt. My plan is to do ontology embedding so that I can embed the high level components of the graph, such as classes, annotation properties, object properties. Then I can store these embedding into a vector database. When the user ask a question in the chatbot, the natural language question is also embeded and then the word embedding is compared and matched with the corresponding ontology embedding to find the related ontology. For example, if the user asks "How many hospitals are there in Amsterdam?", the ontology classes of "building type" and "city" can be matched, because hospital is a building type and Amsterdam is a city. This will serve automatically in the prompt to the backed LLM to generate sparql queries for our question answering system to query the detailed instance data and convert them into natural language answer to our user.
The problem right now is that, I want to build the above pipeline to make the idea work, I need to do ontology embedding and I want to do it with mowl. I have the ontology data(.owl), but I don't have the annotation file for it. It's also not a gene ontology so I'm not sure what alternative can serve as the role of .gaf here for it. I can show you the ontology I use to test the pipeline right now. It's just a small part of the knowledge graph(or more precisely, knowledge graphs). You can find the small ontology in this URI 
http://www.opengis.net/ont/geosparql
I can also attach the owl file I extracted to you. It is not supported to be uploaded in github here. I drop an email to you.

"in the PPI benchmark, we obtain protein--protein interactions from some database and split those interaction as training, validation and testing. Training interactions are added to the Gene Ontology, validation and testing interactions are added to new empty ontologies." Yes but I am not sure how to split the interaction here as training, validation and testing to fit mowl.

**Reply by ferzcam** [reply]
I understand. I do not think you need the validation and testing ontologies. You could just embed the ontology you have directly using any of the methods in mOWL. The are ontology embeddings methods that preserve different aspects of the ontology such as correlation, relational structure or model structure https://arxiv.org/abs/2406.10964. The method you use can also depend on your use case. Also, some methods are inherently unsupervised, for example here https://github.com/bio-ontology-research-group/gda_analysis/blob/main/src/owl2vec/gda.py, and ontology is transformed into a graph, random walks are generated from the graph and then word2vec embeddings are generated from the random walks, there is no validation or testing ontologies here. In principle you could make all the methods unsupervised by not using validation and testing ontologies and just relying on the loss to stop the optimization.

**Reply by 20Bolin** [reply]
Thank you for sharing you new paper! Do you mean I can train the ELEmPPI model without validation and test set? Usually it raises an issue, or are there any other specific methods you refer to in mOWL?

**Reply by 20Bolin** [reply]
> I understand. I do not think you need the validation and testing ontologies. You could just embed the ontology you have directly using any of the methods in mOWL. The are ontology embeddings methods that preserve different aspects of the ontology such as correlation, relational structure or model structure https://arxiv.org/abs/2406.10964. The method you use can also depend on your use case. Also, some methods are inherently unsupervised, for example here https://github.com/bio-ontology-research-group/gda_analysis/blob/main/src/owl2vec/gda.py, and ontology is transformed into a graph, random walks are generated from the graph and then word2vec embeddings are generated from the random walks, there is no validation or testing ontologies here. In principle you could make all the methods unsupervised by not using validation and testing ontologies and just relying on the loss to stop the optimization.

Also when I created a subclass of ELEmbeddings to train a model with opengis.owl, the training seems to go, with some warning messages related to owlapi, but I can't get any saved model in the designated model_filepath.
```
from mowl.models.elembeddings.model import ELEmbeddings
import os
cwd = os.getcwd()
dataset = ds

class OntoEmbedding(ELEmbeddings):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        
    def train(self, validate_every=1000):

        optimizer = th.optim.Adam(self.module.parameters(), lr=self.learning_rate)
        best_loss = float('inf')
        
        prots = [self.class_index_dict[p] for p
                 in self.dataset.evaluation_classes[0].as_str]
        
        for epoch in trange(self.epochs):
            self.module.train()

            train_loss = 0
            loss = 0

            for gci_name, gci_dataset in self.training_datasets.items():
                if len(gci_dataset) == 0:
                    continue

                loss += th.mean(self.module(gci_dataset[:], gci_name))
                if gci_name == "gci2":
                    idxs_for_negs = np.random.choice(prots, size=len(gci_dataset), replace=True)
                    rand_index = th.tensor(idxs_for_negs).to(self.device)
                    data = gci_dataset[:]
                    neg_data = th.cat([data[:, :2], rand_index.unsqueeze(1)], dim=1)
                    loss += th.mean(self.module(neg_data, gci_name, neg=True))
                    
            loss += self.module.regularization_loss()
                    
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            train_loss += loss.detach().item()

            loss = 0
            """
            with th.no_grad():
                self.module.eval()
                valid_loss = 0
                gci2_data = self.validation_datasets["gci2"][:]
                loss = th.mean(self.module(gci2_data, "gci2"))
                valid_loss += loss.detach().item()

            if (epoch + 1) % validate_every == 0:
                if valid_loss < best_loss:
                    best_loss = valid_loss
                    th.save(self.module.state_dict(), self.model_filepath)
                print(f'Epoch {epoch+1}: Train loss: {train_loss} Valid loss: {valid_loss}')
            """
            print(f'Epoch {epoch+1}: Train loss: {train_loss}')
        return 1

    def eval_method(self, data):
        return self.module.gci2_score(data)

ds = CustomDataset(ontology_path="opengis.owl")
dataset = ds

model_filepath=os.path.join(cwd, "Ontomodel.pkl")
model = OntoEmbedding(dataset,
                 embed_dim=30,
                 margin=0.05,
                 reg_norm=1,
                 learning_rate=0.001,
                 epochs=50,
                 batch_size=4096,
                 model_filepath=model_filepath,
                 device='cpu')

model.module.ind_embed = nn.Embedding(num_embeddings=len(dataset.classes), embedding_dim=30)

model.train()
print("Model file path:", model_filepath)
```

I got results in the following snippet:
```
  0%|                                                                                           | 0/50 [00:00<?, ?it/s]INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(13 8)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(23 8)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(24 8)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(25 8)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(26 8)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(14 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(15 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(16 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(17 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(18 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(19 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(20 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(21 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(33 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(34 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(35 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(36 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(37 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(38 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(39 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(40 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(41 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(42 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(43 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(44 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(45 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(46 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(47 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
INFO:root:Reverse translation. Ignoring axiom: RangeAxiom*(48 10)
INFO:root:'uk.ac.manchester.cs.owl.owlapi.OWLObjectPropertyRa' object has no attribute 'getSubClass'
 36%|█████████████████████████████▌                                                    | 18/50 [00:00<00:00, 87.86it/s]
Epoch 1: Train loss: 6.814009189605713
Epoch 2: Train loss: 6.791727066040039
Epoch 3: Train loss: 6.76841402053833
Epoch 4: Train loss: 6.744767189025879
Epoch 5: Train loss: 6.721012115478516
Epoch 6: Train loss: 6.697953701019287
Epoch 7: Train loss: 6.674616813659668
Epoch 8: Train loss: 6.651337623596191
Epoch 9: Train loss: 6.628175258636475
Epoch 10: Train loss: 6.604928970336914
Epoch 11: Train loss: 6.58170747756958
Epoch 12: Train loss: 6.558652400970459
Epoch 13: Train loss: 6.535472393035889
Epoch 14: Train loss: 6.512180328369141
Epoch 15: Train loss: 6.489261150360107
Epoch 16: Train loss: 6.466228008270264
Epoch 17: Train loss: 6.443020343780518
Epoch 18: Train loss: 6.419806480407715
Epoch 19: Train loss: 6.396909236907959
Epoch 20: Train loss: 6.3739848136901855
Epoch 21: Train loss: 6.350953578948975
Epoch 22: Train loss: 6.327829837799072
Epoch 23: Train loss: 6.304683208465576
Epoch 24: Train loss: 6.281750202178955
Epoch 25: Train loss: 6.258724212646484
Epoch 26: Train loss: 6.235835552215576
Epoch 27: Train loss: 6.212948322296143
Epoch 28: Train loss: 6.189960479736328
Epoch 29: Train loss: 6.166983604431152
Epoch 30: Train loss: 6.144233703613281
Epoch 31: Train loss: 6.1213884353637695
Epoch 32: Train loss: 6.098453998565674
Epoch 33: Train loss: 6.0757155418396
Epoch 34: Train loss: 6.052928447723389
Epoch 35: Train loss: 6.030158042907715
Epoch 36: Train loss: 6.007312297821045
Epoch 37: Train loss: 5.984601020812988
Epoch 38: Train loss: 5.961874961853027
Epoch 39: Train loss: 5.939144134521484
Epoch 40: Train loss: 5.916323661804199
Epoch 41: Train loss: 5.89362096786499
Epoch 42: Train loss: 5.8709821701049805
Epoch 43: Train loss: 5.8483123779296875
Epoch 44: Train loss: 5.825550556182861
Epoch 45: Train loss: 5.802829742431641
Epoch 46: Train loss: 5.783956050872803
Epoch 47: Train loss: 5.772589206695557
Epoch 48: Train loss: 5.760953426361084
100%|█████████████████████████████████████████████████████████████████████████████████| 50/50 [00:00<00:00, 142.68it/s]
Epoch 49: Train loss: 5.748967170715332
Epoch 50: Train loss: 5.736679553985596
Model file path: C:\Users\Steven\mOWL\Ontomodel.pkl
```

I checked the [documentation](https://mowl.readthedocs.io/en/latest/model.html) and saw this:
![image](https://github.com/user-attachments/assets/dbfa4c3c-5a1f-4445-a8f6-13ae011ed7a8)
I don't think my model contains a Word2Vec instance, so it would be saved as a [torch.nn.Module](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module). Then I also tried to set the model_filepath as .pt or .pth, but it still doesn't work and I see no saved files, although in my jupyter notebook log it recorded the activity of saving file in this notebook.

How can I actually save the model successfully then?

**Reply by ferzcam** [reply]
Hi, I see you commented the `th.save(...` instruction in your code. I suggest to save the model if the training loss decreases. I was using the validation loss to save the model but you do not have a validation dataset.

**Reply by 20Bolin** [reply]
> Hi, I see you commented the `th.save(...` instruction in your code. I suggest to save the model if the training loss decreases. I was using the validation loss to save the model but you do not have a validation dataset.

Hi thx!!! For the gda method you recommended https://github.com/bio-ontology-research-group/gda_analysis/blob/main/src/owl2vec/gda.py#L45, you mentioned that it was inherently unsupervised. But gda is actually a supervised learning for classification. So there is a little confusion here. But I guess by gda perhaps you mean something else than gaussian discriminant analysis?

**Reply by ferzcam** [reply]
The example you mentioned uses the OWL2Vec* star method (https://arxiv.org/abs/2009.14654) to generate ontology embeddings. The generation of embeddings can be unsupervised. The use case (and the ontologies used) of the example in the link corresponds to the biomedical gene-disease association (GDA) problem.

**Reply by 20Bolin** [reply]
> The example you mentioned uses the OWL2Vec* star method (https://arxiv.org/abs/2009.14654) to generate ontology embeddings. The generation of embeddings can be unsupervised. The use case (and the ontologies used) of the example in the link corresponds to the biomedical gene-disease association (GDA) problem.

Thank you! It helps a lot!!
